From 9842b1f35452eaf8b2ca7252905bc6501a956202 Mon Sep 17 00:00:00 2001
From: Michael Hudson-Doyle <michael.hudson@canonical.com>
Date: Wed, 6 Jan 2021 12:18:24 +1300
Subject: [PATCH] cmd/link: check CGO_CFLAGS for non -g/-I/-O options before
 internal linking

On Ubuntu we are investigating enabling link-time optimization by
default, which means the default CFLAGS will contain arguments that
cause gcc to generate object files cmd/link/internal/loadelf cannot
process. Rather than failing in this situation, scan CGO_CFLAGS in
mustLinkExternal and do not link internally if there is a flag that does
not start with -g, -I, or -O.

CFLAGS can also be injected via #cgo CFLAGS: directives but as use of
any non-standard library cgo modules disables internal linking anyway,
we don't have to worry about that here.

Fixes #43505

Change-Id: Ib083f6daf22617e2e5df67e95d3bc178942328cd
---
 src/cmd/link/internal/ld/config.go | 28 ++++++++++++++++++++++++++++
 1 file changed, 28 insertions(+)

--- a/src/cmd/link/internal/ld/config.go
+++ b/src/cmd/link/internal/ld/config.go
@@ -8,6 +8,8 @@
 	"cmd/internal/sys"
 	"fmt"
 	"internal/buildcfg"
+	"os"
+	"strings"
 )
 
 // A BuildMode indicates the sort of object we are building.
@@ -248,6 +250,29 @@
 	return false, ""
 }
 
+// Some arguments in CGO_CFLAGS can cause the host compiler to
+// generate object files that loadelf cannot handle but arguments
+// starting with any of these values are OK.
+var internalOKCflagPrefixes = []string{
+	"-O",
+	"-g",
+	"-I",
+}
+
+func cflagsOKForInternalLinking() bool {
+outer:
+	for _, flag := range strings.Fields(os.Getenv("CGO_CFLAGS")) {
+		for _, okPrefix := range internalOKCflagPrefixes {
+			if strings.HasPrefix(flag, okPrefix) {
+				continue outer
+			}
+		}
+		return false
+	}
+
+	return true
+}
+
 // determineLinkMode sets ctxt.LinkMode.
 //
 // It is called after flags are processed and inputs are processed,
@@ -272,6 +297,9 @@
 		default:
 			if extNeeded || (iscgo && externalobj) {
 				ctxt.LinkMode = LinkExternal
+			} else if iscgo && !cflagsOKForInternalLinking() {
+				ctxt.LinkMode = LinkExternal
+				extReason = "CGO_CFLAGS contains flag that may inhibit internal linking"
 			} else {
 				ctxt.LinkMode = LinkInternal
 			}
